### PR TITLE
feat: add support for original_dst for windows

### PR DESCRIPTION
**Pull request by keithmattix**
I noticed there wasn't support for original dst in Windows, so I added it. Also add windows gnu target to allow cross compiling

### CODE REVIEW

**Review by Thomasdezeeuw — COMMENTED** [review]
It seems that at least Linux has `SO_ORIGINAL_DST` and `IP6T_SO_ORIGINAL_DST` as well, can we add it for it as well?

**Comment by keithmattix** [comment]
@Thomasdezeeuw it's already added in the sys/unix.rs file: https://github.com/rust-lang/socket2/blob/3a938932829ea6ee3025d2d7a86c7b095c76e6c3/src/sys/unix.rs#L2420

**Comment by Thomasdezeeuw** [comment]
> @Thomasdezeeuw it's already added in the sys/unix.rs file:
> 
> https://github.com/rust-lang/socket2/blob/3a938932829ea6ee3025d2d7a86c7b095c76e6c3/src/sys/unix.rs#L2420

Ah, in that case it should be moved to the `src/socket.rs` file to ensure the API is the same across OS.

**Comment by keithmattix** [comment]
Ok yeah I had that thought but wanted to confirm. Will make the change!

**Comment by keithmattix** [comment]
@Thomasdezeeuw any clue what's going on with the FreeBSD CI? From what I can tell, it looks like it's failing due to the test runners not support ipv6

File: src/socket.rs [hunk 1]
```diff
@@ -27,6 +30,38 @@ use crate::{Domain, Protocol, SockAddr, TcpKeepalive, Type};
...
+    ($fn: ident ( $($arg: expr),* $(,)* ) ) => {{
+        #[allow(unused_unsafe)]
+        let res = unsafe { libc::$fn($($arg, )*) };
+        if res == -1 {
+            Err(std::io::Error::last_os_error())
+        } else {
+            Ok(res)
+        }
+    }};
+}
```
Comment (Thomasdezeeuw) [inline src/socket.rs:63]: Why are these changes needed?

File: src/socket.rs [hunk 2]
```diff
@@ -2205,6 +2240,120 @@ impl Socket {
...
+                        storage.cast(),
+                        len,
+                    ),
+                    PartialEq::eq,
+                    SOCKET_ERROR
+                )
+            })
+        }
+        .map(|(_, addr)| addr)
+    }
```
Comment (Thomasdezeeuw) [inline src/socket.rs:2356]: These functions should be merged with the Unix version, see all the other functions here.

**Review by Thomasdezeeuw — CHANGES_REQUESTED** [review]

File: src/socket.rs (hunk 1 above)
Comment (keithmattix) [reply src/socket.rs:63]: The OS specific syscall macros only existed in the sys module before this PR. The original destination functions utilize them so I duplicated it from sys

File: src/socket.rs (hunk 2 above)
Comment (keithmattix) [reply src/socket.rs:2356]: Ah I see, I misunderstood your original instructions. Will fix

**Comment by keithmattix** [comment]
@Thomasdezeeuw gentle ping for review

File: src/sys/unix.rs [hunk 3]
```diff
@@ -1406,6 +1406,62 @@ pub(crate) const fn to_mreqn(
     }
 }
 
+/// Get the value for the `SO_ORIGINAL_DST` option on this socket.
+///
+/// This value contains the original destination IPv4 address of the connection
+/// redirected using `iptables` `REDIRECT` or `TPROXY`.
```
Comment (Thomasdezeeuw) [inline src/sys/unix.rs:1412]: Nit: all these docs are now hidden, so we might as well remove them or move the to the public method on `Socket`, same for the Windows side.
```suggestion
```

File: src/socket.rs [hunk 4]
```diff
@@ -2205,6 +2205,48 @@ impl Socket {
...
+        )))
+    )]
+    pub fn original_dst(&self) -> io::Result<SockAddr> {
+        sys::original_dst(self.as_raw())
+    }
+
+    /// Get the value for the `IP6T_SO_ORIGINAL_DST` option on this socket.
+    #[cfg(all(
+        feature = "all",
+        any(target_os = "android", target_os = "linux", target_os = "windows",)
```
Comment (Thomasdezeeuw) [inline src/socket.rs:2238]:
```suggestion
        any(target_os = "android", target_os = "linux", target_os = "windows")
```

File: src/socket.rs [hunk 5]
```diff
@@ -2205,6 +2205,48 @@ impl Socket {
...
+    ))]
+    #[cfg_attr(
+        docsrs,
+        doc(cfg(all(
+            feature = "all",
+            any(target_os = "android", target_os = "linux", target_os = "windows",)
```
Comment (Thomasdezeeuw) [inline src/socket.rs:2244]:
```suggestion
            any(target_os = "android", target_os = "linux", target_os = "windows")
```

File: tests/socket.rs [hunk 6]
```diff
@@ -1607,7 +1610,27 @@ fn original_dst() {
 }
 
 #[test]
-#[cfg(all(feature = "all", any(target_os = "android", target_os = "linux")))]
+#[cfg(all(feature = "all", target_os = "windows"))]
```
Comment (Thomasdezeeuw) [inline tests/socket.rs:1613]: The tests should be the same across OS, so please merge it with the `original_dst` above. Same for `original_dst_ipv6`.

File: src/sys/unix.rs [hunk 7]
```diff
@@ -1406,6 +1406,58 @@ pub(crate) const fn to_mreqn(
...
+    feature = "all",
+    any(target_os = "android", target_os = "fuchsia", target_os = "linux")
+))]
+#[cfg_attr(
+    docsrs,
+    doc(cfg(all(
+        feature = "all",
+        any(target_os = "android", target_os = "fuchsia", target_os = "linux")
+    )))
+)]
```
Comment (Thomasdezeeuw) [inline src/sys/unix.rs:1419]:
```suggestion
```

File: src/sys/unix.rs [hunk 8]
```diff
@@ -1406,6 +1406,58 @@ pub(crate) const fn to_mreqn(
...
+
+/// Get the value for the `IP6T_SO_ORIGINAL_DST` option on this socket.
+///
+/// This value contains the original destination IPv6 address of the connection
+/// redirected using `ip6tables` `REDIRECT` or `TPROXY`.
+#[cfg(all(feature = "all", any(target_os = "android", target_os = "linux")))]
+#[cfg_attr(
+    docsrs,
+    doc(cfg(all(feature = "all", any(target_os = "android", target_os = "linux"))))
+)]
```
Comment (Thomasdezeeuw) [inline src/sys/unix.rs:1444]:
```suggestion
```

File: src/sys/windows.rs [hunk 9]
```diff
@@ -857,6 +859,64 @@ pub(crate) fn to_mreqn(
...
+/// Get the value for the `SO_ORIGINAL_DST` option on this socket.
+/// Only valid for sockets in accepting mode.
+///
+/// Note: if using this function in a proxy context, you must query the
+/// redirect records for this socket and set them on the outbound socket
+/// created by your proxy in order for any OS level firewall rules to be
+/// applied. Read more in the Windows bind and connect redirection
+/// [documentation](https://learn.microsoft.com/en-us/windows-hardware/drivers/network/using-bind-or-connect-redirection).
+#[cfg(feature = "all")]
+#[cfg_attr(docsrs, doc(cfg(all(windows, feature = "all"))))]
```
Comment (Thomasdezeeuw) [inline src/sys/windows.rs:871]:
```suggestion
#[cfg(feature = "all")]
```

File: src/sys/windows.rs [hunk 10]
```diff
@@ -857,6 +859,64 @@ pub(crate) fn to_mreqn(
...
+/// Get the value for the `IP6T_SO_ORIGINAL_DST` option on this socket.
+/// Only valid for sockets in accepting mode.
+///
+/// Note: if using this function in a proxy context, you must query the
+/// redirect records for this socket and set them on the outbound socket
+/// created by your proxy in order for any OS level firewall rules to be
+/// applied. Read more in the Windows bind and connect redirection
+/// [documentation](https://learn.microsoft.com/en-us/windows-hardware/drivers/network/using-bind-or-connect-redirection).
+#[cfg(feature = "all")]
+#[cfg_attr(docsrs, doc(cfg(all(windows, feature = "all"))))]
```
Comment (Thomasdezeeuw) [inline src/sys/windows.rs:900]:
```suggestion
#[cfg(feature = "all")]
```

File: tests/socket.rs [hunk 11]
```diff
@@ -42,6 +42,7 @@ use std::num::NonZeroUsize;
 use std::os::unix::io::AsRawFd;
 #[cfg(windows)]
 use std::os::windows::io::AsRawSocket;
+
```
Comment (Thomasdezeeuw) [inline tests/socket.rs:45]:
```suggestion
```

**Review by Thomasdezeeuw — APPROVED** [review]

**Comment by Thomasdezeeuw** [comment]
Thanks @keithmattix
